components/filter/filter.js

```diff
@@ -0,0 +1,270 @@
+		return html`
+			<div slot="header">
+				<div class="d2l-filter-dimension-header">
+					<d2l-button-icon
+						@click="${this._handleDimensionHide}"
+						icon="tier1:chevron-left"
+						text="${this.localize('components.menu-item-return.returnCurrentlyShowing', 'menuTitle', dimensionText)}">
+					</d2l-button-icon>
+					<div class="d2l-filter-dimension-header-text d2l-body-standard">${dimensionText}</div>
+				</div>
+			</div>
+		`;
+	}
+
+	_createSetDimension(dimension) {
+		return html`
+			<d2l-list
+				@d2l-list-selection-change="${this._handleChangeSetDimension}"
+				extend-separators>
+
+				${dimension.values.map(item => html`
+					<d2l-list-item
+						key="${item.key}"
+						selectable
+						?selected="${item.selected}"
+						slim>
+
+						<div class="d2l-filter-dimension-set-value-text">${item.text}</div>
+					</d2l-list-item>
+				`)}
+			</d2l-list>
+		`;
+	}
+
+	_dispatchChangeEvent(eventDetail) {
+		this.dispatchEvent(new CustomEvent('d2l-filter-change', { bubbles: false, composed: false, detail: eventDetail }));
+	}
+
+	_getSlottedNodes(slot) {
+		const nodes = slot.assignedNodes();
+		return nodes.filter((node) => node.nodeType === Node.ELEMENT_NODE && node.tagName.toLowerCase() !== 'template');
+	}
+
+	_handleChangeSetDimension(e) {
+		const dimensionKey = e.composedPath()[0].parentNode.id;
+		const valueKey = e.detail.key;
+		const selected = e.detail.selected;
+
+		// Update the corresponding d2l-filter-dimension-set-value to keep them in sync
+		const dimension = this.querySelector(`d2l-filter-dimension-set[key="${dimensionKey}"`);
+		const slot = dimension.shadowRoot.querySelector('slot');
+		const items = this._getSlottedNodes(slot);
+		const item = items.find(item => item.key === valueKey);
+		item.selected = selected;
+
+		this._dispatchChangeEvent({ dimension: dimensionKey, value: { key: valueKey, selected: selected } });
+	}
+
+	_handleDimensionDataChange(e) {
+		const dimension = this._dimensions.find(dimension => dimension.key === e.detail.dimensionKey);
+		const value = e.detail.valueKey && dimension.values.find(value => value.key === e.detail.valueKey);
+		const toUpdate = value ? value : dimension;
+		const changes = e.detail.changes;
+
+		changes.forEach((newValue, prop) => {
+			if (toUpdate[prop] !== newValue) {
+				toUpdate[prop] = newValue;
+				this.requestUpdate();
+			}
+		});
+	}
+
+	_handleDimensionHide() {
+		this.shadowRoot.querySelector(`#${this._activeDimensionKey}`).hide();
+		this._activeDimensionKey = null;
+	}
+
+	_handleDimensionShowComplete() {
+		const dimension = this.shadowRoot.querySelector(`#${this._activeDimensionKey}`);
+		const focusable = getFirstFocusableDescendant(dimension);
+		if (focusable) {
+			focusable.focus();
+		}
+	}
+
+	_handleDimensionShowStart(e) {
+		this._activeDimensionKey = e.detail.sourceView.id;
+	}
+
+	_handleDropdownClose(e) {
+		this._activeDimensionKey = null;
+		this._stopPropagation(e);
+	}
+
+	_handleSlotChange() {
```

Comment:
Do we still need to do any of this DOM crawling when slots change? Was hoping we could rely solely on the events fired from the dimensions themselves to maintain our state. That might mean firing an even when they're disconnected to remove them.

Reply:
Hmm, not sure I'm following - are you thinking instead of a `slotchange` event which tells us nothing about _what_ nodes changed, we'd just listen for `d2l-filter-dimension-added` and `d2l-filter-dimension-removed` events (and likewise, `d2l-filter-dimension-set` would need to listen for `d2l-filter-dimension-set-value-added` and `d2l-filter-dimension-set-value-removed`, etc)?

Reply:
Yeah exactly, that way you could avoid reaching into the DOM entirely. You may even be able to avoid the initial `*-added` events and just assume if you see a `change` with a key you don't have yet then it's new. But you would need `*-removed` events.

Reply:
Do we know that the order will always be as expected?  As in, Lit will go through in the order they appear and I'll get the events in the same order as the DOM nodes?

Reply:
The event ordering is my biggest worry with this approach... the DOM order but also that the events might fire before the parent is even listening. I think we've had issues with that before. So walking the DOM may be unavoidable.

Reply:
Looks like I'd need to hookup the event listener in `connectedCallback`.  The order is as expected for my little test, but I'm less confident once we get HM calls involved.  And what if someone slotted in another "optional" dimension between two others after the fact, when a HM call completed and they determined they did want that dimension?

The `slotchange` event really _shouldn't_ be firing very often - twice at the beginning, and then never again unless you add or remove dimensions or options.  Adding options is probably the most likely case if loading/paging.  What I _could_ do to try to help that is put each component more in charge of it's own data.  The `dimension-set-value` maintains its `info` object that the `dimension-set` points to, and it maintains its own `info` object that the `_dimensions` array in filter points to.  Then, if something changes, the component updates it's own info and tells filter to `requestUpdate` - that way, the `slotchange` event doesn't cause a dom crawl of _all_ the dimension nodes when an option changes.

Reply:
Right, so maybe just better to leave things as you have them. I'm not super worried about crawling the whole DOM if it only happens rarely, was mostly just trying to avoid _any_ DOM crawling... but seems like that's not worth it given the ordering issues you outlined.

Reply:
The initial dispatch of the event could be delayed if necessary.  For the case where a dimension is inserted afterwards, the handler could determine its index and update the filter accordingly.  However, I think the `slotchange` is probably safer from an ordering/timing perspective.